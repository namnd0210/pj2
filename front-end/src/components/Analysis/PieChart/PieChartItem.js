import React from 'react'

export default function PieChartItem(data) {
  const { style, x, y } = data
  const sum = data.data.reduce((a, b) => a + b.y, 0)
  return (
    <svg>
      <g>
        <circle cx={x} cy={y - 6} r="10" fill={style.fill} />
      </g>
      <text x={x + 25} y={y} style={{ fontSize: "20" }}>
        {data.datum.x}
      </text>
      <text x={x + 250} y={y} style={{ fontSize: "20" }}>
        {data.datum.y}
      </text>
      <text x={x + 25} y={y + 35} style={{ fontSize: "30" }}>
        {Math.round(data.datum.y / sum * 100)}%
      </text>
    </svg >
  )
}
