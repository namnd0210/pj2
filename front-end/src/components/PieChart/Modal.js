import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import _ from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck, faAngleDown } from '@fortawesome/free-solid-svg-icons';

import { fetchPieChartData } from '../../actions';

export default ({ startDate, endDate }) => {
  const { data } = useSelector(state => state.pieChartData)
  const [handleData, setHandleData] = useState(data);
  const dispatch = useDispatch();
  const [modal, setModal] = useState(false);

  const toggle = () => {
    setModal(!modal)
    if(modal) dispatch(fetchPieChartData(startDate, endDate, handleData))
  };

  const onClick = (item, index) => {
    setHandleData([
      ..._.slice(handleData, 0, index),
      { ...item, isActive: !item.isActive },
      ..._.slice(handleData, index + 1)
    ])
  }

  return (
    <div>
      <Button color="success" onClick={toggle} style={{ padding: "10px 20px" }}>+</Button>
      <Modal isOpen={modal} toggle={toggle}>
        <ModalHeader toggle={toggle} charCode={<FontAwesomeIcon icon={faAngleDown} />}></ModalHeader>
        <ModalBody>
          <div>
            {
              _.map(handleData, (item, index) =>
                <div
                  className="row"
                  onClick={() => onClick(item, index)}
                  style={{ cursor: "pointer" }}
                >
                  <div className="col-10">{item.x}</div>
                  {item.isActive &&
                    <div className="col-2">
                      <FontAwesomeIcon icon={faCheck} />
                    </div>
                  }
                </div>
              )
            }
          </div>
        </ModalBody>
        <ModalFooter>
          <Button color="primary" onClick={toggle}>OK</Button>
        </ModalFooter>
      </Modal>
    </div>
  );
}
