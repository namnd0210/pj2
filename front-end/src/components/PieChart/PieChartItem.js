import React from 'react'

export default function PieChartItem(data) {
  const { style, x, y } = data
  const sum = data.data.reduce((a, b) => a + b.y, 0)
  return (
    <svg>
      <g>
        <circle cx={x - 12} cy={y - 6} r="7" fill={style.fill} />
      </g>
      <text x={x} y={y} style={{ fontSize: "20" }}>
        {data.datum.x}
      </text>
      <text x={x + 200} y={y} style={{ fontSize: "20" }}>
        {data.datum.y}
      </text>
      <text x={x} y={y + 35} style={{ fontSize: "30" }}>
        {Math.round(data.datum.y / sum * 100)}%
      </text>
    </svg >
  )
}
