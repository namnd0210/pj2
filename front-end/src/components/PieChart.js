import React, { Component } from 'react'
import {
  VictoryPie, VictoryLegend
} from 'victory'
import { connect } from 'react-redux'

import { PieChartItem } from './PieChartItem'
import { fetchPieChartData } from '../actions'
import { bindActionCreators } from 'redux'

const colorScale = ["#79c1b6", "#a285de"]

class PieChart extends Component {
  componentDidMount() {
    this.props.fetchPieChartData()
  }

  render() {
    const { data } = this.props
    return (
      <svg viewBox="0 0 800 800">
        <VictoryPie
          standalone={false}
          data={data}
          colorScale={colorScale}
          innerRadius={100}
          style={{
            labels: { display: "none" }
          }}
        />
        <VictoryLegend x={420} y={130}
          gutter={20}
          rowGutter={{ top: 0, bottom: 60 }}
          data={data}
          standalone={false}
          dataComponent={<PieChartItem />}
          colorScale={colorScale}
        />
      </svg >
    )
  }
}

const mapStateToProps = (state) => {
  return {
    data: state.pieChartData
  }
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({ fetchPieChartData }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(PieChart)
