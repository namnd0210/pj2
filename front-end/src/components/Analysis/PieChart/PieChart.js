import React, { useEffect, useState } from 'react';
import {
  VictoryPie, VictoryLegend
} from 'victory';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';

import PieChartItem from './PieChartItem';
import { fetchPieChartData } from '../../../actions';
import Loading from '../../Loading';
import Modal from './Modal';

const colorScale = ['#A4036F', '#17B890', '#DA394E', '#FFC253', '#A54657', '#EDF060', '#F0803C'];

function shuffle(a) { // shuffle colors
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}
export default function PieChart({ startDate, endDate }) {
  const { isLoading, data } = useSelector(state => state.pieChartData)
  const dispatch = useDispatch()
  const [handleData, setHandleData] = useState([])

  useEffect(() => {
    dispatch(fetchPieChartData(startDate, endDate));
  }, [startDate, endDate])

  useEffect(() => {
    if (data !== []) {
      setHandleData(_.filter(data, o => o.isActive))
    }
  }, [data])

  return (
    <div>
      <div className="row">
        <h2 className="col-7">Summary Device</h2>
        {!isLoading && <Modal className="col-5" startDate={startDate} endDate={endDate} />}
      </div>
      {isLoading && <Loading color="#15e62a" />}
      {!isLoading &&
        <svg viewBox="0 0 700 700">
          <VictoryPie
            standalone={false}
            data={handleData}
            colorScale={colorScale}
            innerRadius={100}
            style={{
              labels: { display: "none" }
            }}
          />
          <VictoryLegend x={400} y={50}
            gutter={20}
            rowGutter={{ top: 0, bottom: 60 }}
            data={handleData}
            standalone={false}
            dataComponent={<PieChartItem />}
            colorScale={shuffle(colorScale)}
          />
        </svg>}
    </div>
  );
}