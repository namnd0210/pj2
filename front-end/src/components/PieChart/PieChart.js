import React, { useEffect, useState } from 'react';
import {
  VictoryPie, VictoryLegend
} from 'victory';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';

import PieChartItem from './PieChartItem';
import { fetchPieChartData } from '../../actions';
import Loading from '../Loading';
import Modal from './Modal';

const colorScale = ['#a65b5b', '#f2865e', '#f2bea0', '#fad56a', '#f8e4a9', '#788abf'];

export default function PieChart({ startDate, endDate }) {
  const { isLoading, data } = useSelector(state => state.pieChartData)
  const dispatch = useDispatch()
  const [handleData, setHandleData] = useState(data)

  useEffect(() => {
    dispatch(fetchPieChartData(startDate, endDate));
  }, [startDate, endDate])

  useEffect(() => {
    if (data !== []) {
      setHandleData(_.filter(data, o => o.isActive))
    }
  }, [data])

  return (
    <svg viewBox="0 0 700 700">
      <foreignObject x="0" y="0" width="100%" height="50px">
        <div className="row">
          <h2 className="col-7">Summary Device</h2>
          {!isLoading && <Modal className="col-5" startDate={startDate} endDate={endDate} />}
        </div>
      </foreignObject>
      {isLoading && <Loading color="#15e62a" />}
      {!isLoading &&
        <g>
          <VictoryPie
            standalone={false}
            data={handleData}
            colorScale={colorScale}
            innerRadius={100}
            style={{
              labels: { display: "none" }
            }}
          />
          <VictoryLegend x={420} y={50}
            gutter={20}
            rowGutter={{ top: 0, bottom: 60 }}
            data={handleData}
            standalone={false}
            dataComponent={<PieChartItem />}
            colorScale={colorScale}
          />
        </g>}
    </svg>
  );
}