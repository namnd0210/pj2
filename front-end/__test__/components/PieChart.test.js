import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import PieChart from '../../src/components/Analysis/PieChart/PieChart'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('react-redux', () => ({
  useDispatch: () => { },
  useSelector: () => ({})
}));

describe('<PieChart />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(
      <PieChart />,
    );
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  })
  it('Should render and match the snapshot', () => {
    const wrapper = shallow(
      <PieChart />,
    );
    expect(wrapper).toMatchSnapshot();
  })
})