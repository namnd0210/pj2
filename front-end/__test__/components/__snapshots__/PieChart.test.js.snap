// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PieChart /> Should render and match the snapshot 1`] = `ShallowWrapper {}`;
